docs/language/informal/toplevel-inference.md

```diff
@@ -0,0 +1,293 @@
+first inferring methods, and then doing inference for "field like things"
+(static variables, fields, setters, and getters), where we don't allow inference
+for "field like things" to depend on the results of inference for "field like
+things" except in cases where the dependencies are statically visible.
+
+
+Some broad principles for type inference that the language team has agreed to,
+and which this proposal is designed to satisfy:
+* Type inference should be free to fail with an error, rather than being
+  required to always produce some answer
+* It should not be possible for a programmer to observe a declaration as having
+  two different types (e.g. dynamic for recursive uses, but subsequently int).
+  Some consistent answer (or an error) should always be produced.
+* The inference for local variables, toplevel variables, and fields, should
+  either agree or error out.  The same expression should not be inferred
+  differently at different syntactic positions.  It’s ok for an expression to be
+  inferrable at one level but not at another.
+* Obvious types should be inferred
+* Inferred and annotated types should be treated the same
+* Simple intuition for users
+* Efficient to implement
+* Type arguments should be mostly inferred
+
+
+A motivating example for the current strong mode implementation (that also
+serves as a demonstration of its shortcomings) is as follows:
+
+
+```dart
+class A {
+  var x = 3;
+}
+var a = new A();
+var b = a.x;
+```
+
+
+To ensure stability with respect to cycles and different orderings, current
+strong mode does inference in two passes: first on the static variables, and
+secondly on fields (method override based inference is mostly irrelevant for
+this discussion). The implementation ensures that inference is stable, but it
+has several unfortunate consequences, most notably the following.  Inference may
+result in the same variable being implicitly seen at two different types. In the
+example above, if all declarations are in the same library, then b will be
+inferred before x has been inferred, and so b will be inferred to have type
+dynamic.  The same declaration in a separate library will see the type of a.x
+after inference, and hence will be inferred to have type `int`.  This violates
+one of the principles above, and also means that changes to the library
+structure that introduce a cycle may change inference results.
+
+
+## Method inference
+
+
+### Method inference
+
+
+Method inference for a method m behaves as if all method inference for all of
+its supertypes has already been performed.
+
+
+If a method leaves a type off of its signature (parameter type or return type)
+and it does not override or implement anything from a super type, it is an
+error.
+
+
+If a method leaves a type off of its signature (parameter type or return type)
+and it does not override or implement anything from a super type, then the
+omitted types are treated as dynamic.
```

Comment:
Am I missing something, or is the sentence at line 91 contradicting the one at line 86?